File: backend/app/schemas/skill_evaluation.py
from __future__ import annotations

import uuid
from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, field_validator
from pydantic_core import PydanticCustomError

from app.agent_runtime.skill_builder.eval_limits import MAX_SKILL_EVAL_CASES
from app.schemas.skill_builder import JsonValue
from app.services.skill_evaluation_case_limits import (
    SkillEvaluationCaseSizeError,
    validate_evaluation_case_sizes,
)


class SkillEvaluationRunStatus(StrEnum):
    QUEUED = "queued"
    RUNNING = "running"
    GRADING = "grading"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"


class SkillEvaluationSetCreate(BaseModel):
    model_config = ConfigDict(frozen=True)

    name: str = Field(..., min_length=1, max_length=160)
    description: str | None = None
    evals: list[JsonValue] = Field(..., min_length=1, max_length=MAX_SKILL_EVAL_CASES)

    @field_validator("evals")
    @classmethod
    def validate_eval_case_sizes(cls, value: list[JsonValue]) -> list[JsonValue]:
        return _bounded_eval_cases(value)


class SkillEvaluationSetUpdate(BaseModel):
    model_config = ConfigDict(frozen=True)

    name: str | None = Field(default=None, min_length=1, max_length=160)
    description: str | None = None
    evals: list[JsonValue] | None = Field(
        default=None,
        min_length=1,
        max_length=MAX_SKILL_EVAL_CASES,
    )

    @field_validator("evals")
    @classmethod
    def validate_eval_case_sizes(cls, value: list[JsonValue] | None) -> list[JsonValue] | None:
        if value is None:
            return None
        return _bounded_eval_cases(value)


def _bounded_eval_cases(value: list[JsonValue]) -> list[JsonValue]:
    try:
        validate_evaluation_case_sizes(value)
    except SkillEvaluationCaseSizeError as exc:
        raise PydanticCustomError("skill_eval_case_too_large", str(exc)) from exc
    return value


class SkillEvaluationRunEstimate(BaseModel):
    model_config = ConfigDict(frozen=True)

    case_count: int = Field(..., ge=0)
    model_call_count: int = Field(..., ge=0)
    estimated_seconds: int = Field(..., ge=0)
    timeout_seconds: int = Field(..., ge=1)
    estimated_tokens_in: int = Field(default=0, ge=0)
    estimated_tokens_out: int = Field(default=0, ge=0)
    estimated_cost_usd: float = Field(..., ge=0)
    # False → the runner model has no per-token pricing; estimated_cost_usd=0
    # then means "unknown", not "free" (spec §5.2).
    pricing_available: bool = False
    runner_model: str | None = None
    uses_baseline_comparison: bool


class SkillEvaluationRunCreateRequest(BaseModel):
    model_config = ConfigDict(frozen=True)

    # False skips the measured without-arm (2 calls/case instead of 3).
    baseline_comparison: bool = True


class SkillEvaluationRunCancelRequest(BaseModel):
    model_config = ConfigDict(frozen=True)

    reason: str = Field(default="user", max_length=120)


class SkillEvaluationPrepareRequest(BaseModel):
    model_config = ConfigDict(frozen=True)

    allow_llm_generation: bool = True
    force: bool = False


class SkillEvaluationPrepareResponse(BaseModel):
    model_config = ConfigDict(frozen=True)

    status: str
    evaluation_set_id: uuid.UUID | None = None
    source_kind: str
    case_count: int = Field(..., ge=0)
    payload_hash: str | None = None
    reason: str | None = None


class SkillEvaluationRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, frozen=True)

    id: uuid.UUID
    skill_id: uuid.UUID
    evaluation_set_id: uuid.UUID
    status: SkillEvaluationRunStatus
    skill_version: str | None = None
    skill_content_hash: str | None = None
    runner_model: str | None = None
    summary: dict[str, JsonValue] | None = None
    benchmark: dict[str, JsonValue] | None = None
    # Measured LLM usage rollup — {model_calls, tokens_in, tokens_out,
    # cost_usd, measured}. None for legacy/deterministic runs.
    usage: dict[str, JsonValue] | None = None
    case_results: list[JsonValue] | None = None
    error_message: str | None = None
    cancellation_requested_at: datetime | None = None
    cancellation_reason: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None
    created_at: datetime
    updated_at: datetime


class SkillEvaluationSetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True, frozen=True)

    id: uuid.UUID
    skill_id: uuid.UUID
    name: str
    description: str | None = None
    source_kind: str
    evals: list[JsonValue]
    expectations_schema_version: int
    latest_run: SkillEvaluationRunResponse | None = None
    created_at: datetime
    updated_at: datetime

File: frontend/src/lib/types/skill-evaluation.ts
import type { JsonValue } from './json'

export type SkillEvaluationRunStatus =
  | 'queued'
  | 'running'
  | 'grading'
  | 'completed'
  | 'failed'
  | 'cancelled'

export type SkillHealthState =
  | 'ready'
  | 'needs_evaluation'
  | 'needs_rerun'
  | 'needs_credentials'
  | 'evaluation_running'
  | 'evaluation_failed'
  | 'low_confidence'

export type SkillHealthSeverity = 'success' | 'info' | 'warning' | 'error' | 'neutral'

export type SkillHealthSummary = {
  readonly state: SkillHealthState
  readonly label: string
  readonly reason: string
  readonly severity: SkillHealthSeverity
}

export type SkillEvaluationRunEstimate = {
  readonly case_count: number
  readonly model_call_count: number
  readonly estimated_seconds: number
  readonly timeout_seconds: number
  readonly estimated_tokens_in?: number
  readonly estimated_tokens_out?: number
  readonly estimated_cost_usd: number
  /** false → runner 모델 단가 미보유. cost 0은 "무료"가 아니라 "모름". */
  readonly pricing_available?: boolean
  readonly runner_model?: string | null
  readonly uses_baseline_comparison: boolean
}

/** llm-2 실측 usage rollup — 레거시/deterministic 런은 null. */
export type SkillEvaluationRunUsage = {
  readonly measured?: boolean
  readonly model_calls?: number
  readonly tokens_in?: number
  readonly tokens_out?: number
  readonly cost_usd?: number | null
}

export type SkillEvaluationVersionStats = {
  readonly skill_version?: string | null
  readonly content_hash?: string | null
  readonly run_count: number
  readonly latest_pass_rate?: number | null
  readonly avg_pass_rate?: number | null
  readonly latest_pass_rate_delta?: number | null
  readonly latest_measured: boolean
  readonly first_run_at: string
  readonly last_run_at: string
}

export type SkillCaseFeedbackVerdict = 'agree' | 'disagree'

export type SkillCaseFeedback = {
  readonly run_id: string
  readonly case_index: number
  readonly verdict: SkillCaseFeedbackVerdict
  readonly comment?: string | null
  readonly updated_at: string
}

export type SkillCaseFeedbackUpsert = {
  readonly case_index: number
  readonly verdict: SkillCaseFeedbackVerdict
  readonly comment?: string | null
}

export type SkillEvaluationRun = {
  readonly id: string
  readonly skill_id: string
  readonly evaluation_set_id: string
  readonly status: SkillEvaluationRunStatus
  readonly skill_version?: string | null
  readonly skill_content_hash?: string | null
  readonly runner_model?: string | null
  readonly summary?: Readonly<Record<string, JsonValue>> | null
  readonly benchmark?: Readonly<Record<string, JsonValue>> | null
  readonly usage?: SkillEvaluationRunUsage | null
  readonly case_results?: readonly JsonValue[] | null
  readonly error_message?: string | null
  readonly cancellation_requested_at?: string | null
  readonly cancellation_reason?: string | null
  readonly started_at?: string | null
  readonly completed_at?: string | null
  readonly created_at: string
  readonly updated_at: string
}

export type SkillLatestEvaluationSummary = {
  readonly status: SkillEvaluationRunStatus | 'missing' | 'stale' | 'partial' | 'passed'
  readonly latest_run_id?: string | null
  readonly evaluation_set_id?: string | null
  readonly pass_rate?: number | null
  readonly skill_content_hash?: string | null
  readonly created_at?: string | null
  readonly completed_at?: string | null
}

export type SkillEvaluationSet = {
  readonly id: string
  readonly skill_id: string
  readonly name: string
  readonly description?: string | null
  readonly source_kind: string
  readonly evals: readonly JsonValue[]
  readonly expectations_schema_version: number
  readonly latest_run?: SkillEvaluationRun | null
  readonly created_at: string
  readonly updated_at: string
}

export type SkillEvaluationSetCreate = {
  readonly name: string
  readonly description?: string | null
  readonly evals: readonly JsonValue[]
}

export type SkillEvaluationSetUpdate = {
  readonly name?: string | null
  readonly description?: string | null
  readonly evals?: readonly JsonValue[] | null
}

export type SkillEvaluationRunCancelRequest = {
  readonly reason?: string
}
